test/pkg/integration/workspace.go

```diff
@@ -266,42 +248,71 @@ func LaunchWorkspaceFromContextURL(ctx context.Context, contextURL string, usern
 		return nil, nil, xerrors.Errorf("cannot start workspace: %q", err)
 	}
 
-	nfo, err := server.GetWorkspace(ctx, resp.CreatedWorkspaceID)
+	t.Logf("attemp to get the workspace information: %s", resp.CreatedWorkspaceID)
+	wi, err := server.GetWorkspace(ctx, resp.CreatedWorkspaceID)
 	if err != nil {
 		return nil, nil, xerrors.Errorf("cannot get workspace: %q", err)
 	}
-	if nfo.LatestInstance == nil {
+	if wi.LatestInstance == nil {
 		return nil, nil, xerrors.Errorf("CreateWorkspace did not start the workspace")
 	}
+	t.Logf("got the workspace information: %s", wi.Workspace.ID)
 
 	// GetWorkspace might receive an instance before we seen the first event
 	// from ws-manager, in which case IdeURL is not set
-	nfo.LatestInstance.IdeURL = resp.WorkspaceURL
-
-	stopWs := func(waitForStop bool) {
-		sctx, scancel := context.WithTimeout(ctx, perCallTimeout)
-		_ = server.StopWorkspace(sctx, resp.CreatedWorkspaceID)
-		scancel()
+	wi.LatestInstance.IdeURL = resp.WorkspaceURL
 
-		if waitForStop {
-			_, _ = WaitForWorkspaceStop(ctx, api, nfo.LatestInstance.ID)
-		}
-	}
+	stopWs := stopWsF(t, wi.LatestInstance.ID, api)
 	defer func() {
 		if err != nil {
 			stopWs(false)
 		}
 	}()
-	// it.t.Logf("created workspace: workspaceID=%s url=%s", resp.CreatedWorkspaceID, resp.WorkspaceURL)
 
-	_, err = WaitForWorkspaceStart(ctx, nfo.LatestInstance.ID, api)
+	t.Log("wait for workspace to be fully up and running")
+	_, err = WaitForWorkspaceStart(ctx, wi.LatestInstance.ID, api)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("cannot start workspace: %q", err)
+		return nil, nil, xerrors.Errorf("failed to wait for the workspace to start up: %w", err)
 	}
+	t.Log("successful launch of the workspace")
+
+	return wi, stopWs, nil
+}
 
-	// it.t.Logf("workspace is running: instanceID=%s", nfo.LatestInstance.ID)
+func stopWsF(t *testing.T, instanceID string, api *ComponentAPI) stopWorkspaceFunc {
+	return func(waitForStop bool) (*wsmanapi.WorkspaceStatus, error) {
+		sctx, scancel := context.WithTimeout(context.Background(), perCallTimeout)
+		defer scancel()
 
-	return nfo, stopWs, nil
+		var err error
+		for {
+			t.Logf("attemp to delete the workspace: %s", instanceID)
+			err = DeleteWorkspace(sctx, api, instanceID)
+			if st, ok := status.FromError(err); ok && st.Code() == codes.Unavailable {
+				t.Logf("got Unavailable during when try to delete the workspace: %v", st)
+				time.Sleep(5 * time.Second)
+				continue
+			} else if err != nil {
+				return nil, err
+			}
+			break
+		}
+		if waitForStop {
```

Comment:
```suggestion
                if !waitForStop {
```

https://github.com/gitpod-io/go-style-guide/blob/master/style.md#reduce-nesting

Reply:
Good point